sql: `insert values not provided for all table columns` probably due to rolling back a schema change

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainGist [failed](https://mesolite.cluster.engflow.com/invocations/default/88dc33b4-2377-400d-ae58-2a8edaf374ef?testReportRun=6&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on release-24.3.21-rc @ [d343b53a21d34412715ed228d74ca4f114cf147c](https://github.com/cockroachdb/cockroach/commits/d343b53a21d34412715ed228d74ca4f114cf147c):


```
        						(oid(764737415251875100:::INT8::INT8)::OID),
        						(1748857052:::OID)
        				)
        					AS tab_367 (col_506)
        		),
        	with_47 (col_507)
        		AS (
        			SELECT
        				*
        			FROM
        				(VALUES (1266565537:::REGCLASS), (regclass(e'[C=h``\x0f\x03t&':::STRING::STRING)::REGCLASS))
        					AS tab_368 (col_507)
        		),
        	with_48 (col_508)
        		AS (
        			SELECT
        				7.986362842021567822E+33:::DECIMAL AS col_508
        			FROM
        				defaultdb.public.seed AS tab_369
        			WHERE
        				tab_369._bool
        		)
        SELECT
        	(-28486):::INT8 AS col_509,
        	(-1426764660):::INT8 AS col_510,
        	'1997-10-31 23:13:09.000074':::TIMESTAMP AS col_511,
        	NULL AS col_512,
        	0.1934438998392940658:::DECIMAL AS col_513,
        	'\x5de1bb':::BYTES AS col_514,
        	'f9488f62-1608-438d-b6b3-5f9d0d9cd59b':::UUID AS col_515,
        	(
        		SELECT
        			'hello':::greeting AS col_516
        		FROM
        			defaultdb.public.seed AS tab_370, defaultdb.public.seed@seed__int8__float8__date_idx AS tab_371
        		WHERE
        			tab_370._bool
        		ORDER BY
        			tab_371._date NULLS LAST
        		LIMIT
        			1:::INT8
        	)
        		AS col_517
        FROM
        	with_47 AS cte_ref_20, with_46 AS cte_ref_21
        ORDER BY
        	cte_ref_20.col_507 NULLS LAST, cte_ref_20.col_507 DESC
        LIMIT
        	77:::INT8;
    --- FAIL: TestExplainGist/main (5.30s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainGist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-55090

--- a/pkg/sql/schemachanger/dml_injection_test.go
+++ b/pkg/sql/schemachanger/dml_injection_test.go
@@ -8,6 +8,7 @@ package schemachanger_test
 import (
 	"context"
 	"fmt"
+	"regexp"
 	"slices"
 	"strings"
 	"sync/atomic"
@@ -144,6 +145,11 @@ type testCase struct {
 	// Optional: If you want a query to run at each stage, you can include it here.
 	// We don't evaluate the results; we simply assert that the query executes without errors.
 	query string
+	// expectedDMLErr, when non-empty, is a regexp that matches DML errors
+	// (INSERT/DELETE/UPDATE against the original columns) that are tolerated
+	// during schema change stages. Errors not matching this pattern still
+	// fail the test.
+	expectedDMLErr string
 }
 
 // Captures testCase before t.Parallel is called.
@@ -192,6 +198,14 @@ func TestAlterTableDMLInjection(t *testing.T) {
 			setup:        []string{"CREATE SEQUENCE seq"},
 			schemaChange: "ALTER TABLE tbl ADD COLUMN new_col INT NOT NULL DEFAULT nextval('seq')",
 			expectedErr:  "cannot evaluate scalar expressions containing sequence operations in this context",
+		},
+		{
+			// The REGCLASS default requires a catalog lookup during backfill, which
+			// fails. During rollback the column transitions WRITE_ONLY → DELETE_ONLY
+			// → ABSENT; concurrent DML must succeed throughout.
+			desc:         "add column default regclass",
+			schemaChange: "ALTER TABLE tbl ADD COLUMN new_col INT NOT NULL DEFAULT ('tbl'::REGCLASS)::INT",
+			expectedErr:  "cannot evaluate scalar expressions using table lookups in this context",
 		},
 		{
 			desc:         "add column default serial rowid",
@@ -299,9 +313,10 @@ func TestAlterTableDMLInjection(t *testing.T) {
 			schemaChange: "ALTER TABLE tbl DROP COLUMN new_col",
 		},
 		{
-			desc:         "add column virtual NOT NULL",
-			schemaChange: "ALTER TABLE tbl ADD COLUMN new_col TEXT NOT NULL AS (NULL::TEXT) VIRTUAL",
-			expectedErr:  "validation of column \"new_col\" NOT NULL failed on row: insert_phase_ordinal='pre-schema-change', operation_phase_ordinal='n/a', operation='n/a', val=1, new_col=NULL",
+			desc:           "add column virtual NOT NULL",
+			schemaChange:   "ALTER TABLE tbl ADD COLUMN new_col TEXT NOT NULL AS (NULL::TEXT) VIRTUAL",
+			expectedErr:    "validation of column \"new_col\" NOT NULL failed on row: insert_phase_ordinal='pre-schema-change', operation_phase_ordinal='n/a', operation='n/a', val=1, new_col=NULL",
+			expectedDMLErr: "NOT NULL",
 		},
 		{
 			desc:         "add column virtual",
@@ -600,6 +615,10 @@ func TestAlterTableDMLInjection(t *testing.T) {
 			currentPOForBackFill.Store(phaseOrdinal{})
 			var beforeBackfillCallbackDone, afterBackfillCallbackDone atomic.Bool
 			var backfillCallback func(isAfter bool)
+			var expectedDMLRe *regexp.Regexp
+			if tc.expectedDMLErr != "" {
+				expectedDMLRe = regexp.MustCompile(tc.expectedDMLErr)
+			}
 			testCluster := serverutils.StartCluster(t, 1, base.TestClusterArgs{
 				ServerArgs: base.TestServerArgs{
 					Knobs: base.TestingKnobs{
@@ -638,8 +657,46 @@ func TestAlterTableDMLInjection(t *testing.T) {
 									return nil
 								}
 
-								// Cannot verify DML injection for unsupported schema changes.
+								// For schema changes expected to error we cannot run the
+								// full DML-verification matrix, but we still inject basic
+								// DML to ensure concurrent operations do not get unexpected
+								// errors during failure/rollback.
 								if tc.expectedErr != "" {
+									s := p.Stages[stageIdx]
+									isRollback := p.InRollback || p.CurrentState.InRollback
+									stageDesc := fmt.Sprintf(
+										"%s:%d(rollback=%t)",
+										s.Phase, s.Ordinal, isRollback,
+									)
+									reportDMLErr := func(op string, err error) {
+										if err == nil {
+											return
+										}
+										if expectedDMLRe != nil && expectedDMLRe.MatchString(err.Error()) {
+											t.Logf("%s: %s failed (expected): %v", stageDesc, op, err)
+											return
+										}
+										// Log as t.Errorf to fail the test, but continue running
+										// this goroutine to prevent hangs.
+										t.Errorf("%s: %s failed: %v", stageDesc, op, err)
+									}
+									dmlKey := fmt.Sprintf("err-dml-%s", stageDesc)
+									_, insertErr := sqlDB.DB.ExecContext(
+										ctx, insert, dmlKey, na, na, valInitial,
+									)
+									reportDMLErr("INSERT", insertErr)
+									_, deleteErr := sqlDB.DB.ExecContext(
+										ctx,
+										`DELETE FROM tbl WHERE insert_phase_ordinal = $1`,
+										dmlKey,
+									)
+									reportDMLErr("DELETE", deleteErr)
+									_, updateErr := sqlDB.DB.ExecContext(
+										ctx,
+										`UPDATE tbl SET val = val + 1 WHERE insert_phase_ordinal = $1`,
+										dmlKey,
+									)
+									reportDMLErr("UPDATE", updateErr)
 									return nil
 								}
 

